src/language-markdown/printer-markdown.js

```diff
@@ -540,15 +547,104 @@ function getAncestorNode(path, typeOrTypes) {
+ *
+ * @param {*} path current position in nodes tree
+ * @returns {boolean} `true` if Space is tend to be inserted between these types of letters, `false` otherwise.
+ */
+function isSentenceUseCJDividingSpace(path) {
+  const sentenceNode = getAncestorNode(path, "sentence");
+  if (sentenceNode.isCJSpacingUsing !== undefined) {
+    return sentenceNode.isCJSpacingUsing;
+  }
+
+  const cjNonCJKSpacingStatistics = {
+    " ": 0,
+    "": 0,
+  };
+  for (let i = 0; i < sentenceNode.children.length; ++i) {
+    const node = sentenceNode.children[i];
+    if (node.type === "whitespace") {
+      switch (node.value) {
+        case " ":
+        case "": {
+          const previous = sentenceNode.children[i - 1];
+          const next = sentenceNode.children[i + 1];
+          if (
+            !(
+              (previous?.kind === "cj-letter" && next?.kind === "non-cjk") ||
+              (previous?.kind === "non-cjk" && next?.kind === "cj-letter")
+            )
+          ) {
+            continue;
+          }
+          ++cjNonCJKSpacingStatistics[node.value];
+          continue;
+        }
+      }
+    }
+  }
+  // Injects a property to cache the result.
+  sentenceNode.isCJSpacingUsing =
+    cjNonCJKSpacingStatistics[" "] > cjNonCJKSpacingStatistics[""];
+  return sentenceNode.isCJSpacingUsing;
+}
+
+/**
+ * Looks for 1st `:::` (MUST be followed by a Space) from 2nd `:::`.
+ *
+ * ```
+ * ::: foo
+ * bar
+ * :::
+ * ```
+ *
+ * @param {*} path current position in nodes tree
+ * @param {*} mark what to look for
+ * @returns `true` if a `mark` followed by Space (U+0020) is found before `path`, `false` otherwise.
+ */
+function isCorrespondingMarkFollowedBySpaceBefore(path, mark) {
+  const sentenceNode = getAncestorNode(path, "sentence");
+  if (
+    sentenceNode.children
+      .slice(0, getPenultimate(path.stack))
+      .some(
+        (node, i, array) =>
+          node.value === mark &&
+          array[i + 1]?.type === "whitespace" &&
+          array[i + 1]?.value === " "
+      )
+  ) {
+    return true;
+  }
+  return false;
+}
+
+function printLine(path, value, options, adjacentNodes) {
   if (options.proseWrap === "preserve" && value === "\n") {
     return hardline;
   }
 
   const isBreakable =
     options.proseWrap === "always" &&
     !getAncestorNode(path, SINGLE_LINE_NODE_TYPES);
-  return value !== ""
+  // Chinese and Japanese does not use U+0020 Space to dive words, so U+00A0 No-break space must not be replaced with it.
+  // Behavior in other languages will not be changed because there are too much things to consider. (PR welcome)
+  const canLineBreakBeConvertedToSpace = !(
+    value === "\n" &&
+    typeof adjacentNodes === "object" &&
+    (adjacentNodes?.previous?.kind === "cj-letter" ||
+      adjacentNodes?.next?.kind === "cj-letter") &&
+    // we wonder if there are other marks to be considered.
+    (adjacentNodes.next.value !== ":::" ||
```

Comment:
Can you add tests for `:::`?

Reply:
What is `:::`?

Reply:
Some markdown processor supports `:::` syntax https://docusaurus.io/docs/next/markdown-features/admonitions
I'm not sure how this affects to CJ letters